### PR TITLE
Use protocolVersion in MessageSerializer

### PR DESCRIPTION
**Pull request by erasmospunk**
This change moves protocolVersion from Message to MessageSerializer.
The main reason is to support parsing transactions as non-segwit.
This is impossible to do otherwise, as MessageSerializer#makeTransaction
does not accept a protocolVersion parameter.
Additionally this change makes the MessageSerializer similar to the
Stream in BitcoinCore, where the protocol version can be modified.

I marked the API as beta so we can discuss a potentially better implementation.

### CODE REVIEW

File: core/src/main/java/org/bitcoinj/core/BitcoinSerializer.java [hunk 1]
```diff
@@ -378,4 +404,19 @@ public BitcoinPacketHeader(ByteBuffer in) throws ProtocolException, BufferUnderf
             cursor += 4;
         }
     }
+
```
Comment (schildbach) [inline core/src/main/java/org/bitcoinj/core/BitcoinSerializer.java]: I would say adding equals/hashCode to BitcoinSerializer could go into a separate commit?

File: core/src/main/java/org/bitcoinj/core/Message.java [hunk 2]
```diff
@@ -342,7 +346,18 @@ protected long readVarInt(int offset) throws ProtocolException {
...
         }
     }
-    
+
+    protected byte readByte() throws ProtocolException {
+        checkReadLength(1);
+        return payload[cursor++];
+    }
+
+    private void checkReadLength(int length) throws ProtocolException {
```
Comment (schildbach) [inline core/src/main/java/org/bitcoinj/core/Message.java]: This "extract method" refactoring could go into a separate commit, couldn't it?

**Comment by schildbach** [comment]
> I marked the API as beta so we can discuss a potentially better implementation.

Is there something you expect to change specifically?

**Comment by erasmospunk** [comment]
I think no, but I wanted to hear some feedback about the API.

(irrelevant but @schildbach how to contact you personally? I sent you an email but I'm not sure you are using that one)

File: core/src/main/java/org/bitcoinj/core/ChildMessage.java [hunk 3]
```diff
@@ -43,7 +43,7 @@ public ChildMessage(NetworkParameters params, byte[] payload, int offset, int pr
         super(params, payload, offset, protocolVersion);
     }
 
-    public ChildMessage(NetworkParameters params, byte[] payload, int offset, int protocolVersion, Message parent, MessageSerializer setSerializer, int length) throws ProtocolException {
+    public ChildMessage(NetworkParameters params, byte[] payload, int offset, int protocolVersion, @Nullable Message parent, MessageSerializer setSerializer, int length) throws ProtocolException {
```
[the remark this answers is not in the document]
Comment (schildbach) [reply core/src/main/java/org/bitcoinj/core/ChildMessage.java]: I went ahead and fixed this via 73583776a6ef86e3d624047199747d43d38994f1.

File: core/src/main/java/org/bitcoinj/core/Message.java (hunk 2 above)
Comment (schildbach) [reply core/src/main/java/org/bitcoinj/core/Message.java]: I just noticed you extracted the `checkReadLength()` method in order to add a  `readByte()` method. But what are you using this new method for? I don't see any references.

File: core/src/main/java/org/bitcoinj/core/BitcoinSerializer.java (hunk 1 above)
Comment (schildbach) [reply core/src/main/java/org/bitcoinj/core/BitcoinSerializer.java]: I added a variant of this via 3cbb10334edae29b86dbc08686bf61c487851061.

**Comment by schildbach** [comment]
I merged the rest of this PR, minus the unused methods. Thanks for your contribution, I think this was a much needed refactoring!
